
# <span style="font-variant: small-caps"> Big Data </span>

## Assignment 1
zuzan

---

Download the 311 Service Requests from 2010 to present from https://nycopendata.socrata.com/Social-Services/311-Service-Requests-from-2010-to-Present/erm2-nwe9

**Task 1** Working with plain csv file, find:
* the most common complaints,
* the most common complaints in each borough,
* the agencies with the most common complaints.

Measure the times needed to complete the queries.

**Task 2** Now convert the data from the csv file into an SQL database, then complete the above tasks again.
* What are the times needed for the queries?
* What is the total time needed to complete the task, i.e. to convert the data and find the queries?
* Compare the execution times for both SQLite and Mysql databases.

**Task 3** How would you improve the efficiency of searching for queries on a standalone computer? Implement your ideas and present the results (even in the case of no performance gains!).

-----

### Part I &ndash; answering questions using plain csv and databases

### *plain csv*

In [1]:
import pandas as pd
from dask import dataframe as dd
import time

I am using Dask library which enables to work on large datasets automatically computing queries in blocks. I limit the dataset to only four columns I need to answer questions.

In [2]:
dfd = dd.read_csv(
    '311_Service_Requests_from_2010_to_Present.csv',
    blocksize=64000000, # = 64 Mb chunks
    usecols=['Complaint Type', 'Borough', 'Agency', 'Agency Name']
)

# this takes no time, the data is not really stored in the memory yet

In [3]:
dfd.head()

Agency                                        Agency Name Complaint Type  \
0    HPD  Department of Housing Preservation and Develop...     WATER LEAK   
1    HPD  Department of Housing Preservation and Develop...     WATER LEAK   
2    HPD  Department of Housing Preservation and Develop...     WATER LEAK   
3    DOB                            Department of Buildings   Building/Use   
4    HPD  Department of Housing Preservation and Develop...    DOOR/WINDOW   

     Borough  
0   BROOKLYN  
1   BROOKLYN  
2     QUEENS  
3     QUEENS  
4  MANHATTAN

In [4]:
len(dfd)

22420583

 * the most common complaints

In [11]:
start_time = time.time()
complaint_types = dfd[['Complaint Type']].groupby(['Complaint Type']).size().compute()
complaint_types = complaint_types.reset_index(name='counts').sort_values('counts', ascending=False)[:50]   # 50 most common compl
print("--- %s seconds ---" % (time.time() - start_time))

--- 65.72635507583618 seconds ---


In [12]:
complaint_types

Complaint Type   counts
75                   Noise - Residential  1843661
48                        HEAT/HOT WATER  1276333
103                     Street Condition   975493
57                       Illegal Parking   946167
11                      Blocked Driveway   940933
104               Street Light Condition   924963
298                              HEATING   887869
85                              PLUMBING   712158
127                         Water System   654475
46         General Construction/Plumbing   504705
426                 GENERAL CONSTRUCTION   500863
71                                 Noise   486059
76               Noise - Street/Sidewalk   478465
115                 UNSANITARY CONDITION   453599
113             Traffic Signal Condition   440224
96                                 Sewer   364987
320                      PAINT - PLASTER   361257
72                    Noise - Commercial   354913
33                      Dirty Conditions   350524
84                         PAINT/PLASTER   347093
184  Request Large Bulky Item Collection   326335
35                              ELECTRIC   307820
93                  Sanitation Condition   301471
28                          Damaged Tree   292315
90                                Rodent   275210
16                          Building/Use   274072
68     Missed Collection (All Materials)   266647
428                             NONCONST   260890
97                    Sidewalk Condition   256055
77                       Noise - Vehicle   221539
32                     Derelict Vehicles   214207
31                      Derelict Vehicle   207814
26                           DOOR/WINDOW   206049
124                           WATER LEAK   194352
47                              Graffiti   174250
15                     Broken Muni Meter   172119
23                    Consumer Complaint   170816
111                       Taxi Complaint   169968
70                      New Tree Request   156472
45                               GENERAL   151938
83               Overgrown Tree/Branches   145536
40                       FLOORING/STAIRS   137692
53            Homeless Person Assistance   121209
66               Maintenance or Facility   118542
314             Benefit Card Replacement   113957
0                              APPLIANCE   113299
38                              Elevator   109022
268       DOF Property - Reduction Issue   108501
91         Root/Sewer/Sidewalk Condition    92068
286                                SCRIE    87627

 * the most common complaints in each borough

In [23]:
start_time = time.time()
complaint_borough = dfd[['Borough', 'Complaint Type']].groupby(['Borough','Complaint Type']).size().compute()
complaint_borough = complaint_borough.reset_index(name='counts').groupby('Borough').apply(lambda x: x.nlargest(5, 'counts')).reset_index(drop=True)
print("--- %s seconds ---" % (time.time() - start_time))

--- 70.41128540039062 seconds ---


In [24]:
complaint_borough

Borough            Complaint Type  counts
0           BRONX       Noise - Residential  457533
1           BRONX            HEAT/HOT WATER  416433
2           BRONX                   HEATING  195284
3           BRONX    Street Light Condition  187785
4           BRONX                  PLUMBING  180193
5        BROOKLYN       Noise - Residential  549669
6        BROOKLYN            HEAT/HOT WATER  388566
7        BROOKLYN          Blocked Driveway  344549
8        BROOKLYN           Illegal Parking  343657
9        BROOKLYN          Street Condition  270739
10      MANHATTAN       Noise - Residential  412765
11      MANHATTAN            HEAT/HOT WATER  288820
12      MANHATTAN                     Noise  233413
13      MANHATTAN   Noise - Street/Sidewalk  193778
14      MANHATTAN          Street Condition  171712
15         QUEENS          Blocked Driveway  383365
16         QUEENS       Noise - Residential  365792
17         QUEENS          Street Condition  296232
18         QUEENS           Illegal Parking  289045
19         QUEENS    Street Light Condition  284641
20  STATEN ISLAND          Street Condition  122778
21  STATEN ISLAND    Street Light Condition   60760
22  STATEN ISLAND           Illegal Parking   57017
23  STATEN ISLAND              Water System   56463
24  STATEN ISLAND       Noise - Residential   55044
25    Unspecified                   HEATING  282916
26    Unspecified      GENERAL CONSTRUCTION  150277
27    Unspecified                  PLUMBING  132297
28    Unspecified  Benefit Card Replacement  113956
29    Unspecified           PAINT - PLASTER  110072

 * the agencies with the most common complaints

In [30]:
common_comp = complaint_types['Complaint Type'].to_list() # 50 most common complaints
common_comp

['Noise - Residential',
 'HEAT/HOT WATER',
 'Street Condition',
 'Illegal Parking',
 'Blocked Driveway',
 'Street Light Condition',
 'HEATING',
 'PLUMBING',
 'Water System',
 'General Construction/Plumbing',
 'GENERAL CONSTRUCTION',
 'Noise',
 'Noise - Street/Sidewalk',
 'UNSANITARY CONDITION',
 'Traffic Signal Condition',
 'Sewer',
 'PAINT - PLASTER',
 'Noise - Commercial',
 'Dirty Conditions',
 'PAINT/PLASTER',
 'Request Large Bulky Item Collection',
 'ELECTRIC',
 'Sanitation Condition',
 'Damaged Tree',
 'Rodent',
 'Building/Use',
 'Missed Collection (All Materials)',
 'NONCONST',
 'Sidewalk Condition',
 'Noise - Vehicle',
 'Derelict Vehicles',
 'Derelict Vehicle',
 'DOOR/WINDOW',
 'WATER LEAK',
 'Graffiti',
 'Broken Muni Meter',
 'Consumer Complaint',
 'Taxi Complaint',
 'New Tree Request',
 'GENERAL',
 'Overgrown Tree/Branches',
 'FLOORING/STAIRS',
 'Homeless Person Assistance',
 'Maintenance or Facility',
 'Benefit Card Replacement',
 'APPLIANCE',
 'Elevator',
 'DOF Property - Re

In [51]:
start_time = time.time()
agencies = dfd[dfd['Complaint Type'].isin(common_comp)]['Agency'].unique().compute()
print("--- %s seconds ---" % (time.time() - start_time))

--- 68.11693239212036 seconds ---


In [52]:
agencies

0       HPD
1       DOB
2       DPR
3       DOT
4      DSNY
5       DEP
6     DOHMH
7      NYPD
8       DHS
9       TLC
10      DCA
11    3-1-1
12    DOITT
13      DOF
14     DFTA
15      HRA
16      ACS
Name: Agency, dtype: object

### *SQLite*

In [41]:
import pandas as pd
import sqlite3
from pandas.io import sql
import time
from IPython.display import display, HTML

Converting csv to SQLite database table

In [42]:
connex = sqlite3.connect("A1.db")  
cur = connex.cursor()  
columns = ['Complaint Type', 'Borough', 'Agency', 'Agency Name']

In [43]:
# use only to create if doesn't exist!

start_time = time.time()
for chunk in pd.read_csv("311_Service_Requests_from_2010_to_Present.csv", chunksize=1000000, usecols=columns):
    chunk.to_sql(name="requests", con=connex, if_exists="append", index=False)
print("--- %s seconds ---" % (time.time() - start_time))

C:\Users\zuzan\Anaconda3\lib\site-packages\pandas\core\generic.py:2663: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  method=method,


--- 156.79635381698608 seconds ---


Reading from SQLite database

In [44]:
start_time = time.time()
sql = "SELECT COUNT(*) FROM requests;" 
display(pd.read_sql_query(sql, connex))
print("--- %s seconds ---" % (time.time() - start_time))   # just checking size of the table

COUNT(*)
0  22420583

--- 3.061711311340332 seconds ---


 * the most common complaints

In [45]:
start_time = time.time()
sql = "SELECT [Complaint Type], COUNT(*) FROM requests GROUP BY [Complaint Type];"  
complaint_types1 = pd.read_sql_query(sql, connex)
complaint_types1 = complaint_types1.sort_values('COUNT(*)', ascending=False)[:50]
print("--- %s seconds ---" % (time.time() - start_time))

--- 23.386937618255615 seconds ---


In [46]:
complaint_types1

Complaint Type  COUNT(*)
301                  Noise - Residential   1843661
190                       HEAT/HOT WATER   1276333
372                     Street Condition    975493
216                      Illegal Parking    946167
78                      Blocked Driveway    940933
373               Street Light Condition    924963
191                              HEATING    887869
318                             PLUMBING    712158
415                         Water System    654475
184        General Construction/Plumbing    504705
181                 GENERAL CONSTRUCTION    500863
296                                Noise    486059
302              Noise - Street/Sidewalk    478465
397                 UNSANITARY CONDITION    453599
391             Traffic Signal Condition    440224
357                                Sewer    364987
316                      PAINT - PLASTER    361257
297                   Noise - Commercial    354913
140                     Dirty Conditions    350524
317                        PAINT/PLASTER    347093
340  Request Large Bulky Item Collection    326335
148                             ELECTRIC    307820
352                 Sanitation Condition    301471
132                         Damaged Tree    292315
343                               Rodent    275210
87                          Building/Use    274072
287    Missed Collection (All Materials)    266647
292                             NONCONST    260890
358                   Sidewalk Condition    256055
303                      Noise - Vehicle    221539
139                    Derelict Vehicles    214207
138                     Derelict Vehicle    207814
127                          DOOR/WINDOW    206049
410                           WATER LEAK    194352
186                             Graffiti    174250
83                     Broken Muni Meter    172119
99                    Consumer Complaint    170816
384                       Taxi Complaint    169968
294                     New Tree Request    156472
180                              GENERAL    151938
315              Overgrown Tree/Branches    145536
163                      FLOORING/STAIRS    137692
208           Homeless Person Assistance    121209
234              Maintenance or Facility    118542
74              Benefit Card Replacement    113957
45                             APPLIANCE    113299
155                             Elevator    109022
123       DOF Property - Reduction Issue    108501
344        Root/Sewer/Sidewalk Condition     92068
346                                SCRIE     87627

 * the most common complaints in each borough

In [47]:
start_time = time.time()
sql = "SELECT Borough, [Complaint Type], COUNT(*) FROM requests GROUP BY Borough, [Complaint Type];"   
complaint_borough1 = pd.read_sql_query(sql, connex)
complaint_borough1 = complaint_borough1.groupby('Borough').apply(lambda x: x.nlargest(5, 'COUNT(*)')).reset_index(drop=True)
print("--- %s seconds ---" % (time.time() - start_time))

--- 50.58084011077881 seconds ---


In [48]:
complaint_borough1

Borough            Complaint Type  COUNT(*)
0           BRONX       Noise - Residential    457533
1           BRONX            HEAT/HOT WATER    416433
2           BRONX                   HEATING    195284
3           BRONX    Street Light Condition    187785
4           BRONX                  PLUMBING    180193
5        BROOKLYN       Noise - Residential    549669
6        BROOKLYN            HEAT/HOT WATER    388566
7        BROOKLYN          Blocked Driveway    344549
8        BROOKLYN           Illegal Parking    343657
9        BROOKLYN          Street Condition    270739
10      MANHATTAN       Noise - Residential    412765
11      MANHATTAN            HEAT/HOT WATER    288820
12      MANHATTAN                     Noise    233413
13      MANHATTAN   Noise - Street/Sidewalk    193778
14      MANHATTAN          Street Condition    171712
15         QUEENS          Blocked Driveway    383365
16         QUEENS       Noise - Residential    365792
17         QUEENS          Street Condition    296232
18         QUEENS           Illegal Parking    289045
19         QUEENS    Street Light Condition    284641
20  STATEN ISLAND          Street Condition    122778
21  STATEN ISLAND    Street Light Condition     60760
22  STATEN ISLAND           Illegal Parking     57017
23  STATEN ISLAND              Water System     56463
24  STATEN ISLAND       Noise - Residential     55044
25    Unspecified                   HEATING    282916
26    Unspecified      GENERAL CONSTRUCTION    150277
27    Unspecified                  PLUMBING    132297
28    Unspecified  Benefit Card Replacement    113956
29    Unspecified           PAINT - PLASTER    110072

 * the agencies with the most common complaints

In [50]:
common_comp = complaint_types1['Complaint Type'].to_list()
common_comp

['Noise - Residential',
 'HEAT/HOT WATER',
 'Street Condition',
 'Illegal Parking',
 'Blocked Driveway',
 'Street Light Condition',
 'HEATING',
 'PLUMBING',
 'Water System',
 'General Construction/Plumbing',
 'GENERAL CONSTRUCTION',
 'Noise',
 'Noise - Street/Sidewalk',
 'UNSANITARY CONDITION',
 'Traffic Signal Condition',
 'Sewer',
 'PAINT - PLASTER',
 'Noise - Commercial',
 'Dirty Conditions',
 'PAINT/PLASTER',
 'Request Large Bulky Item Collection',
 'ELECTRIC',
 'Sanitation Condition',
 'Damaged Tree',
 'Rodent',
 'Building/Use',
 'Missed Collection (All Materials)',
 'NONCONST',
 'Sidewalk Condition',
 'Noise - Vehicle',
 'Derelict Vehicles',
 'Derelict Vehicle',
 'DOOR/WINDOW',
 'WATER LEAK',
 'Graffiti',
 'Broken Muni Meter',
 'Consumer Complaint',
 'Taxi Complaint',
 'New Tree Request',
 'GENERAL',
 'Overgrown Tree/Branches',
 'FLOORING/STAIRS',
 'Homeless Person Assistance',
 'Maintenance or Facility',
 'Benefit Card Replacement',
 'APPLIANCE',
 'Elevator',
 'DOF Property - Re

In [53]:
start_time = time.time()
sql = "SELECT DISTINCT Agency FROM requests WHERE [Complaint Type] IN ({seq});".format(
    seq=','.join(['?']*len(common_comp)))  
agencies1 = pd.read_sql_query(sql, connex, params=tuple(common_comp))
print("--- %s seconds ---" % (time.time() - start_time))

--- 12.170240879058838 seconds ---


In [54]:
agencies1

Agency
0     HPD
1     DOB
2     DPR
3     DOT
4    DSNY
5     DEP
6   DOHMH
7    NYPD
8     DHS
9     TLC
10    DCA
11  3-1-1
12  DOITT
13    DOF
14   DFTA
15    HRA
16    ACS

In [55]:
connex.commit()
connex.close()

### *MySQL*

In [56]:
import pandas as pd
import mysql.connector
import time

import pymysql
import sys
from sqlalchemy import create_engine

In [57]:
connex = mysql.connector.connect(host='localhost',
    user='root',
    passwd='   ')
cur = connex.cursor()

In [61]:
cur.execute("CREATE DATABASE A1")

In [62]:
cur.execute("SHOW DATABASES")

for x in cur:
    print(x)

('information_schema',)
('a1',)
('biblioteka',)
('gwiazdy',)
('mysql',)
('performance_schema',)
('sklep',)
('spa',)
('sys',)
('zakupy',)


In [63]:
columns = ['Complaint Type', 'Borough', 'Agency', 'Agency Name']

In [67]:
# use only to create!
start_time = time.time()
engine = create_engine('mysql+mysqldb://root:{password}@localhost/a1', echo = False)

for chunk in pd.read_csv("311_Service_Requests_from_2010_to_Present.csv", chunksize=1000000, usecols=columns):
    chunk.to_sql(name = 'requests', con = engine, if_exists = 'append', index = False)
print("--- %s seconds ---" % (time.time() - start_time))

--- 694.7147965431213 seconds ---


In [68]:
connex = mysql.connector.connect(host='localhost',
    user='root',
    passwd='   ',
    database='A1')
cur = connex.cursor()

In [69]:
cur.execute('SHOW COLUMNS FROM requests;')

In [70]:
cur.fetchall()

[('Agency', 'text', 'YES', '', None, ''),
 ('Agency Name', 'text', 'YES', '', None, ''),
 ('Complaint Type', 'text', 'YES', '', None, ''),
 ('Borough', 'text', 'YES', '', None, '')]

Querying on text data types consumes a lot of memory and is slow, so I firstly changed column types.

In [71]:
start_time = time.time()
cur.execute('ALTER TABLE requests MODIFY `Agency` VARCHAR(8)')
cur.execute('ALTER TABLE requests MODIFY `Borough` VARCHAR(16)')
cur.execute('ALTER TABLE requests MODIFY `Complaint Type` VARCHAR(64)')
print("--- %s seconds ---" % (time.time() - start_time))

--- 699.1692228317261 seconds ---


In [72]:
start_time = time.time()
sql = "SELECT COUNT(*) FROM requests;" 
display(pd.read_sql_query(sql, connex))
print("--- %s seconds ---" % (time.time() - start_time))   # just checking size of the table

COUNT(*)
0  22420583

--- 12.781598329544067 seconds ---


In [73]:
cur.execute(""" SELECT
  TABLE_NAME AS `requests`,
  ROUND((DATA_LENGTH + INDEX_LENGTH) / 1024 / 1024) AS `Size (MB)`
FROM
  information_schema.TABLES
WHERE
  TABLE_SCHEMA = "a1"
ORDER BY
  (DATA_LENGTH + INDEX_LENGTH)
DESC; """)

In [74]:
cur.fetchall()  # mysql database is still larger than sqlite (1647 mb)

[('requests', Decimal('2078'))]

 * the most common complaints

In [76]:
start_time = time.time()
sql = "SELECT `Complaint Type`, COUNT(*) FROM requests GROUP BY `Complaint Type`;"  
complaint_types2 = pd.read_sql_query(sql, connex)
complaint_types2 = complaint_types2.sort_values('COUNT(*)', ascending=False)[:50]
print("--- %s seconds ---" % (time.time() - start_time))

--- 55.19643783569336 seconds ---


In [77]:
complaint_types2

Complaint Type  COUNT(*)
304                  Noise - Residential   1843661
199                       HEAT/HOT WATER   1276333
375                     Street Condition    975493
222                      Illegal Parking    946167
79                      Blocked Driveway    940933
376               Street Light Condition    924963
200                              HEATING    887869
329                             PLUMBING    763804
417                         Water System    654475
187        General Construction/Plumbing    504705
186                 GENERAL CONSTRUCTION    500863
299                                Noise    486059
305              Noise - Street/Sidewalk    478465
405                 UNSANITARY CONDITION    459099
395             Traffic Signal Condition    440224
357                                Sewer    364987
321                      PAINT - PLASTER    361257
300                   Noise - Commercial    354913
121                     Dirty Conditions    350524
322                        PAINT/PLASTER    347093
345  Request Large Bulky Item Collection    326335
151                             ELECTRIC    307821
351                 Sanitation Condition    301471
105                         Damaged Tree    292315
348                               Rodent    275210
88                          Building/Use    274072
291    Missed Collection (All Materials)    266647
310                             NONCONST    260890
359                   Sidewalk Condition    256055
306                      Noise - Vehicle    221539
116                    Derelict Vehicles    214207
115                     Derelict Vehicle    207814
142                          DOOR/WINDOW    206049
414                           WATER LEAK    194353
189                             Graffiti    174250
84                     Broken Muni Meter    172119
101                   Consumer Complaint    170816
388                       Taxi Complaint    169968
297                     New Tree Request    156472
185                              GENERAL    153101
320              Overgrown Tree/Branches    145536
174                      FLOORING/STAIRS    137692
211           Homeless Person Assistance    121209
238              Maintenance or Facility    118542
155                             Elevator    115784
74              Benefit Card Replacement    113957
68                             APPLIANCE    113303
137       DOF Property - Reduction Issue    108501
349        Root/Sewer/Sidewalk Condition     92068
354                                SCRIE     87627

 * the most common complaints in each borough

In [78]:
start_time = time.time()
sql = "SELECT Borough, `Complaint Type`, COUNT(*) FROM requests GROUP BY Borough, `Complaint Type`;"   
complaint_borough2 = pd.read_sql_query(sql, connex)
complaint_borough2 = complaint_borough2.groupby('Borough').apply(lambda x: x.nlargest(5, 'COUNT(*)')).reset_index(drop=True)
print("--- %s seconds ---" % (time.time() - start_time))

--- 71.66827869415283 seconds ---


In [79]:
complaint_borough2

Borough            Complaint Type  COUNT(*)
0           BRONX       Noise - Residential    457533
1           BRONX            HEAT/HOT WATER    416433
2           BRONX                   HEATING    195284
3           BRONX    Street Light Condition    187785
4           BRONX                  PLUMBING    187375
5        BROOKLYN       Noise - Residential    549669
6        BROOKLYN            HEAT/HOT WATER    388566
7        BROOKLYN          Blocked Driveway    344549
8        BROOKLYN           Illegal Parking    343657
9        BROOKLYN          Street Condition    270739
10      MANHATTAN       Noise - Residential    412765
11      MANHATTAN            HEAT/HOT WATER    288820
12      MANHATTAN                     Noise    233413
13      MANHATTAN   Noise - Street/Sidewalk    193778
14      MANHATTAN          Street Condition    171712
15         QUEENS          Blocked Driveway    383365
16         QUEENS       Noise - Residential    365792
17         QUEENS          Street Condition    296232
18         QUEENS           Illegal Parking    289045
19         QUEENS    Street Light Condition    284641
20  STATEN ISLAND          Street Condition    122778
21  STATEN ISLAND    Street Light Condition     60760
22  STATEN ISLAND           Illegal Parking     57017
23  STATEN ISLAND              Water System     56463
24  STATEN ISLAND       Noise - Residential     55044
25    Unspecified                   HEATING    282916
26    Unspecified      GENERAL CONSTRUCTION    150277
27    Unspecified                  PLUMBING    132299
28    Unspecified  Benefit Card Replacement    113956
29    Unspecified           PAINT - PLASTER    110072

 * the agencies with the most common complaints

In [80]:
common_comp = complaint_types2['Complaint Type'].to_list()
common_comp

['Noise - Residential',
 'HEAT/HOT WATER',
 'Street Condition',
 'Illegal Parking',
 'Blocked Driveway',
 'Street Light Condition',
 'HEATING',
 'PLUMBING',
 'Water System',
 'General Construction/Plumbing',
 'GENERAL CONSTRUCTION',
 'Noise',
 'Noise - Street/Sidewalk',
 'UNSANITARY CONDITION',
 'Traffic Signal Condition',
 'Sewer',
 'PAINT - PLASTER',
 'Noise - Commercial',
 'Dirty Conditions',
 'PAINT/PLASTER',
 'Request Large Bulky Item Collection',
 'ELECTRIC',
 'Sanitation Condition',
 'Damaged Tree',
 'Rodent',
 'Building/Use',
 'Missed Collection (All Materials)',
 'NONCONST',
 'Sidewalk Condition',
 'Noise - Vehicle',
 'Derelict Vehicles',
 'Derelict Vehicle',
 'DOOR/WINDOW',
 'WATER LEAK',
 'Graffiti',
 'Broken Muni Meter',
 'Consumer Complaint',
 'Taxi Complaint',
 'New Tree Request',
 'GENERAL',
 'Overgrown Tree/Branches',
 'FLOORING/STAIRS',
 'Homeless Person Assistance',
 'Maintenance or Facility',
 'Elevator',
 'Benefit Card Replacement',
 'APPLIANCE',
 'DOF Property - Re

In [81]:
start_time = time.time()
sql = "SELECT DISTINCT Agency FROM requests WHERE `Complaint Type` IN ({seq});".format(
    seq=','.join(['%s']*len(common_comp)))  
agencies2 = pd.read_sql_query(sql, connex, params=tuple(common_comp))
print("--- %s seconds ---" % (time.time() - start_time))

--- 32.801512479782104 seconds ---


In [82]:
agencies2

Agency
0     HPD
1     DOB
2     DPR
3     DOT
4    DSNY
5     DEP
6   DOHMH
7    NYPD
8     DHS
9     TLC
10    DCA
11  3-1-1
12  DOITT
13    DOF
14   DFTA
15    HRA
16    ACS

In [83]:
connex.commit()
connex.close()

### Part II &ndash; comparison of time needed to answer questions

Time (in seconds) needed to prepare the table and to query:

|               | plain csv     | SQLite  | MySQL   |
| ------------- |--------------:| -------:| -------:|
| preparation   |        ~0     | 156.796 | 694.715 
|               |               |         |+ 699.169|
| query 1       | 65.726        |  23.387 |  55.196 |
| query 2       | 70.411        |  50.581 |  71.668 |
| query 3       | 68.117	    |  12.170 |  32.802 |
| ---           |---            | ---     | ---     |
|**sum queries**|**204.255**    |**86.138**|**159.666**|


The fastest for querying turned out to be SQLite. However, if I consider also the time needed to create a database from csv file, the most efficient for this short task was using Dask and operating on plain csv. Creating MySQL database took more than 10 minutes. Querying without caring for data types in columns took too much time, so I decided to spend these another 10 minutes to set 'lighter' types for the data once, and then query faster. Still the time was not very satisfying and taking into account the whole preparation, using MySQL for solving this task was the worst option.

__How to improve the efficiency of searching for queries on a standalone computer?__

 * including only necessary columns
 * dask - useful library for working with large csv
 * taking care for data types in columns
 * for a wider work with such a huge table, I would choose to create an SQLite database once and then answer questions using it